LLPLabelProducer/plugins/LLPLabelProducer.cc

```diff
@@ -50,6 +60,72 @@ class LLPLabelProducer:
             {
             }
         };
+        
+        static void gatherVisibleDecayProducts(std::vector<const reco::Candidate*>& list, const reco::Candidate* genParticle)
+        {
+            if (genParticle->numberOfDaughters()==0) 
+            {
+                int absId = std::abs(genParticle->pdgId());
+                if (absId!=12 and absId!=14 and absId!=16 and absId>6 and absId<100000)
+                {
+                    list.push_back(genParticle);
+                }
+            }
+            else
+            {
+                for (size_t d=0; d< genParticle->numberOfDaughters(); ++d)
+                {
+                    gatherVisibleDecayProducts(list, genParticle->daughter(d));
+                }
+            }
+        }
+        
+        static reco::Candidate::Point getVertex(const reco::Candidate* p)
+        {
+            const auto genParticle = dynamic_cast<const reco::GenParticle*>(p);
+            if (genParticle) return genParticle->vertex();
+            const auto packedGenParticle = dynamic_cast<const pat::PackedGenParticle*>(p);
+            if (packedGenParticle) return packedGenParticle->vertex();
+            return reco::Candidate::Point(0,0,0);             
+        }
+        
+        static const reco::Candidate* findRecoConstituent(const reco::Candidate* genParticle, const std::vector<const reco::Candidate*>& constituents)
+        {
+            double minDR2 = 1000000;
+            const reco::Candidate* match = nullptr;
+            for (const auto& constituent: constituents)
+            {
+                double ptRel = std::fabs(constituent->pt()/genParticle->pt()-1.);
+                if (ptRel>0.5) continue; //prevent random match
```

Comment:
Shouldn't this be checked if it's not too restrictive? Naively wouldn't deltaR matching be enough?

Reply:
the pT match is done to exclude potential random matches to other particles